File: sdk-js/src/source-part.utils.ts
import type { Part } from "./types.ts";

/**
 * For providers that do not support source parts,
 * we translate them to compatible parts such as Text.
 * Inner source parts are flattened.
 */
export function getCompatiblePartsWithoutSourceParts(parts: Part[]): Part[] {
  return parts
    .map((part) => {
      if (part.type === "source") {
        return getCompatiblePartsWithoutSourceParts(part.content);
      }
      return part;
    })
    .flat();
}

File: sdk-go/sourcepartutils.go
package llmsdk

// [internal] For providers that do not support source parts,
// we translate them to compatible parts such as Text.
// Inner Source parts are flattened.
func GetCompatiblePartsWithoutSourceParts(parts []Part) []Part {
	var result []Part
	for _, part := range parts {
		if part.SourcePart != nil {
			result = append(result, GetCompatiblePartsWithoutSourceParts(part.SourcePart.Content)...)
		} else {
			result = append(result, part)
		}
	}
	return result
}
